src/cargo/ops/cargo_fetch.rs

```diff
@@ -57,6 +58,21 @@ pub fn fetch<'a>(
             .map(|(id, _deps)| id);
         deps_to_fetch.extend(deps);
     }
+
+    // If -Zbuild-std was passed, download dependencies for the standard library.
+    if let Some(crates) = &config.cli_unstable().build_std {
+        standard_lib::ensure_target_passed(&build_config)?;
+
+        // Fetch the dependencies for `test` just in case
+        // (cargo could later decide to build libtest even if it wasn't specified in the flags)
+        let mut crates = crates.clone();
+        standard_lib::ensure_libtest_present(&mut crates);
```

Comment:
Sorry but this sort of feels like you're not really trying to understand my comments and you're just blindly following the first thing that seems to satisfy what I've been mentioning.

To try again, though, what I'm commenting on is the duplication of responsibilities between here and `cargo build`. Duplicated code is naturally difficult to maintain so the general goal with refactorings like this is to ensure that everything stays centralized. This clearly needs to be accounted for in `cargo fetch` but I believe this can be done in a way that doesn't duplicate "don't forget to clone `crates` and optionally add to it".

For example this could be bundled into something that looks like:

```rust
// in standard_lib.rs

pub fn std_crates(config: &Config, tests_present: bool) -> Option<Vec<String>> {
    // ...
}
```

which internally does the clone-and-add logic as necessary.

Unfortunately `-Zbuild-std` is not finished, though, so it's often in need of refactorings to keep it up to date and organized internally. For example as the feature is filled out to go into more commands it's frequently apparent that a previous design needs to be modified. For example the guard requiring `--target` should probably be moved to `BuildConfig::new` to centralize it rather than having a separate `ensure_target_passed` function that consumers of `-Zbuild-std` have to remember to call. Additionally processing of the crates and request could probably happen in a more centralized location instead of accessing `-Zbuild-std` raw configuration input each time.

On balance build-std is an unstable feature and as a contributor to the feature you shouldn't be expected to refactor all of the internals, but at the same time small refactorings to ensure that `-Zbuild-std` remains as bug-free as we can and as close to "ready-ish in the future to stabilize" is what we're shooting for. A big part of that to me is learning, as support is filled in, how the current support needs to be extended rather than simply bolting things on and hoping someone in the future comes through and cleans it all up.